src/xrpld/overlay/detail/Slot.cpp

```diff
@@ -0,0 +1,655 @@
+#include <xrpl/beast/container/aged_unordered_map.h>
+#include <xrpl/beast/utility/Journal.h>
+#include <xrpl/beast/utility/PropertyStream.h>
+#include <xrpl/protocol/PublicKey.h>
+
+#include <chrono>
+#include <optional>
+
+namespace ripple {
+namespace reduce_relay {
+
+void
+Slot::deleteIdlePeer(PublicKey const& validator)
+{
+    using namespace std::chrono;
+    auto now = clock_.now();
+    for (auto it = peers_.begin(); it != peers_.end();)
+    {
+        auto& peer = it->second;
+        auto id = it->first;
+        ++it;
+        if (now - peer.lastMessage > IDLED)
+        {
+            JLOG(journal_.trace())
+                << "deleteIdlePeer: " << Slice(validator) << " " << id
+                << " idled "
+                << duration_cast<seconds>(now - peer.lastMessage).count()
+                << " selected " << (peer.state == PeerState::Selected);
+            deletePeer(validator, id, false);
+        }
+    }
+}
+
+void
+Slot::update(
+    PublicKey const& validator,
+    Peer::id_t id,
+    ignored_squelch_callback callback)
+{
+    using namespace std::chrono;
+    auto const now = clock_.now();
+    auto it = peers_.find(id);
+    // First message from this peer
+    if (it == peers_.end())
+    {
+        JLOG(journal_.trace())
+            << "update: adding peer " << Slice(validator) << " " << id;
+        peers_.emplace(std::make_pair(
+            id,
+            PeerInfo{
+                .state = PeerState::Counting,
+                .count = 0,
+                .expire = now,
+                .lastMessage = now,
+                .timesSelected = 0}));
+        initCounting();
+        return;
+    }
+    // Message from a peer with expired squelch
+    if (it->second.state == PeerState::Squelched && now > it->second.expire)
+    {
+        JLOG(journal_.trace())
+            << "update: squelch expired " << Slice(validator) << " " << id;
+        it->second.state = PeerState::Counting;
+        it->second.lastMessage = now;
+        initCounting();
+        return;
+    }
+
+    auto& peer = it->second;
+
+    JLOG(journal_.trace())
+        << "update: existing peer " << Slice(validator) << " " << id
+        << " slot state " << static_cast<int>(state_) << " peer state "
+        << static_cast<int>(peer.state) << " count " << peer.count << " last "
+        << duration_cast<milliseconds>(now - peer.lastMessage).count()
+        << " pool " << considered_.size() << " threshold " << reachedThreshold_;
+
+    peer.lastMessage = now;
+
+    // report if we received a message from a squelched peer
+    if (peer.state == PeerState::Squelched)
+        callback();
+
+    if (state_ != SlotState::Counting || peer.state == PeerState::Squelched)
+        return;
+
+    if (++peer.count > MIN_MESSAGE_THRESHOLD)
+        considered_.insert(id);
+    if (peer.count == (MAX_MESSAGE_THRESHOLD + 1))
+        ++reachedThreshold_;
+
+    if (now - lastSelected_ > 2 * MAX_UNSQUELCH_EXPIRE_DEFAULT)
```

Comment:
This is defensive block of code. But if it ever happens then we stuck in this state since `lastSelected_` is updated once the peers are selected below. This should never really happen since a slot is deleted if `now - lastSelected_ > MAX_UNSQUELCH_EXPIRE_DEFAULT)`. My question is if `lastSelected_` should be set to `now` if above is true or should we make this an error and delete this slot?

Reply:
So basically the time of the most recent message? I think we just should delete the slot in this case since. This scenario can't happen.

Reply:
I think it can happen. Imagine we have a slot for untrusted validator that stops sending traffic. We will very quickly learn that it's offline, but it'll take an hour until the slot is deleted. Though, and I already started working on this, for untrusted validators, if they fall under the threshold of minimum peers, they slot will be deleted and the validator squelched.